index.js

```diff
@@ -7,27 +7,73 @@
 
 'use strict';
 
-module.exports = function(prefix) {
-  var debug = require('debug');
-
-  return function(app) {
-    this.define('debug', function(name) {
-      var fn = debug(namespace(prefix, name));
-      return function() {
-        return fn.apply(debug, arguments);
-      };
+/**
+ * Plugin that adds `debug` method to the instance
+ * and `namespaces` are added to this method.
+ *
+ * ```js
+ * app.debug()('debugging on main namespace')
+ * //=> base debugging on main namespace
+ *
+ * app.debug.one = app.debug('one')
+ * app.debug.one('testing %s, whatever', 'foo, bar')
+ * //=> base:one testing foo, bar, whatever
+ *
+ * app.debug.mix = app.debug('two', 123 'three:four', 'five')
+ * app.debug.mix('okey, %s awesome', 'this is')
+ * //=> base:two:three:four:five okey, this is awesome
+ *
+ * app.debug.helper('is one of internal namespaces added')
+ * //=> base:helper is one of internal namespaces added
+ * ```
+ *
+ * @name   debug
+ * @param  {Array} `namespaces`
+ * @return {Function}
+ */
+
+module.exports = function debugPlugin(namespaces) {
+  namespaces = Array.isArray(namespaces) ? namespaces : [namespaces];
+
+  return function plugin(app) {
+    app.define('_debugPrefix', app._namespace);
+    app.define('debug', function debug() {
+      return debugFactory.apply(app, arguments);
     });
+    if (namespaces.length) {
+      var len = namespaces.length;
+      var i = 0;
+
+      while (i < len) {
+        var ns = namespaces[i++];
+        app.debug[ns] = debugFactory.call(app, ns);
+        app.debug[ns].color = app.debug.color;
+      }
+    }
   };
 };
 
-function namespace(prefix, subname) {
-  var args = [].slice.call(arguments).filter(Boolean);
-  var name = 'base';
-  if (args.length === 1) {
-    name += ':' + args[0];
-  }
-  if (args.length === 2) {
-    name += ':' + args.join(':');
+function debugFactory() {
+  var app = this;
+  var args = [].slice.call(arguments);
+  var segs = [];
+  var len = args.length;
+  var i = 0;
+
+  while (i < len) {
+    var val = args[i++];
+    if (typeof val === 'string') {
+      segs.push.apply(segs, val.split(':'));
+    }
   }
-  return name;
+
+  return function debug() {
+    var fn = require('debug');
+    var debugSuffix = segs.join(':');
+    var debugNamespace = app._debugPrefix + ':' + debugSuffix;
+    app.define('_debugNamespace', debugNamespace);
+    app.define('_debugSuffix', debugSuffix);
```

Comment:
Hmm, nope. You're not right. It should. They are dynamic, that's the idea of `this.debug` method, will be used for composing.

Example

```js
app.debug.foo = app.debug('foo')
app.debug.foo('debugging msg')
```
They will be set only when `app.debug('foo')` is called, not each time where `app.debug.foo` is called. And it can be in another way, they are dynamic, they depends on what you pass to `app.debug`.